File: backend/app/schemas/batch.py
"""Batch processing schemas for document upload and processing."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class FileResult(BaseModel):
    """Result of processing a single file in a batch."""
    filename: str
    status: str = Field(..., description="completed, failed, processing, pending")
    text: str | None = None
    summary: str | None = None
    key_points: str | None = None
    error: str | None = None
    processing_time_seconds: float | None = None
    document_id: str | None = None


class BatchUploadResponse(BaseModel):
    """Response after submitting a batch upload."""
    batch_id: str
    total_files: int
    status: str = Field(default="processing", description="pending, processing, completed, failed")
    message: str


class BatchStatusResponse(BaseModel):
    """Status of a batch processing job."""
    batch_id: str
    status: str = Field(..., description="pending, processing, completed, failed")
    total_files: int
    completed_files: int
    failed_files: int
    progress_percent: float = Field(..., ge=0, le=100)
    mode: str = Field(default="analyze")
    started_at: str
    completed_at: str | None = None
    results: list[FileResult] = []


class BatchDownloadResponse(BaseModel):
    """Placeholder for batch download info (actual endpoint returns a file)."""
    batch_id: str
    filename: str
    download_url: str


class DocumentItem(BaseModel):
    """A single document in the user's processing history."""
    model_config = ConfigDict(from_attributes=True)

    id: str
    filename: str
    original_filename: str
    file_type: str
    file_size: int
    status: str
    summary: str | None = None
    error_message: str | None = None
    upload_time: datetime
    batch_id: str | None = None


class DocumentHistoryResponse(BaseModel):
    """Paginated list of user's processed documents."""
    documents: list[DocumentItem]
    total: int
    page: int = 1
    page_size: int = 20


class BatchContractReviewResponse(BaseModel):
    """Response after submitting a batch contract review."""
    batch_id: str
    total_files: int
    status: str = Field(default="processing")
    message: str

File: frontend/src/api/batch.ts
import service from './index'

export interface FileResult {
  filename: string
  status: string
  text?: string
  summary?: string
  key_points?: string
  error?: string
  processing_time_seconds?: number
  document_id?: string
}

export interface BatchUploadResponse {
  batch_id: string
  total_files: number
  status: string
  message: string
}

export interface BatchStatusResponse {
  batch_id: string
  status: string
  total_files: number
  completed_files: number
  failed_files: number
  progress_percent: number
  mode: string
  started_at: string
  completed_at: string | null
  results: FileResult[]
}

export interface BatchResultsResponse extends BatchStatusResponse {
  results: FileResult[]
}

export interface DocumentItem {
  id: string
  filename: string
  original_filename: string
  file_type: string
  file_size: number
  status: string
  summary: string | null
  error_message: string | null
  upload_time: string
  batch_id: string | null
}

export interface DocumentHistoryResponse {
  documents: DocumentItem[]
  total: number
  page: number
  page_size: number
}

export const batchApi = {
  uploadBatch(files: File[], mode: string = 'analyze') {
    const formData = new FormData()
    files.forEach((file) => {
      formData.append('files', file)
    })
    formData.append('mode', mode)
    return service.post<BatchUploadResponse>('/document/batch-upload', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    })
  },

  getBatchStatus(batchId: string) {
    return service.get<BatchStatusResponse>(`/document/batch-status/${batchId}`)
  },

  getBatchResults(batchId: string) {
    return service.get<BatchResultsResponse>(`/document/batch-results/${batchId}`)
  },

  downloadBatchResults(batchId: string) {
    return service.get(`/document/batch-download/${batchId}`, {
      responseType: 'blob',
    })
  },

  getDocumentHistory(page: number = 1, pageSize: number = 20) {
    return service.get<DocumentHistoryResponse>('/document/history', {
      params: { page, page_size: pageSize },
    })
  },

  getDocument(documentId: string) {
    return service.get(`/document/${documentId}`)
  },
}
